akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala

```diff
@@ -133,29 +168,29 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
             )
           case immutable.Seq(Failure(e)) =>
             log.warning("Failed to parse http2-settings header in upgrade [{}], continuing with HTTP/1.1", e.getMessage)
-            handler(req)
+            Source.single(req).via(handler).runWith(Sink.head)
           // A server MUST NOT upgrade the connection to HTTP/2 if this header field
           // is not present or if more than one is present
           case _ =>
             log.debug("Invalid upgrade request (http2-settings header missing or repeated)")
-            handler(req)
+            Source.single(req).via(handler).runWith(Sink.head)
         }
 
       case _ =>
-        handler(req)
+        Source.single(req).via(handler).runWith(Sink.head)
```

Comment:
As you said, that probably needs an optimization. Otherwise, HTTP/1 performance will take quite a hit when we need one materialization per request on a persistent connection.